lib/THC/THCCachingAllocator.cpp

```diff
@@ -0,0 +1,282 @@
+#include "THCCachingAllocator.h"
+
+#include <cuda_runtime_api.h>
+#include <mutex>
+#include <set>
+#include <unordered_map>
+#include <map>
+#include <memory>
+
+//
+// Yet another caching allocator for CUDA device allocations.
+//
+// - Allocations are associated with a stream. Once freed, blocks can be
+//   re-allocated on the same stream, but not on any other stream.
+// - The allocator attempts to find the smallest cached block that will fit the
+//   requested size. If the block is larger than the requested size, it may be
+//   split. If no block is found, the allocator will delegate to cudaMalloc.
+// - If the cudaMalloc fails, the allocator will free all cached blocks that
+//   are not split and retry the allocation.
+// - Large (>1MB) and small allocation requestss are handled separately. Large
+//   allocation requests can be filled by a cudaMalloc call of the exact size.
+//   Small requests will allocate and split a 1MB buffer, if necessary.
+
+namespace {
+
+const int ONE_MB = 1048576;
+
+struct Block {
+  int           device;     // gpu
+  cudaStream_t  stream;     // allocation stream
+  size_t        size;       // block size in bytes
+  uintptr_t     ptr;        // memory address
+  bool          allocated;  // in-use flag
+  Block*        prev;       // prev block if split from a larger allocation
+  Block*        next;       // next block if split from a larger allocation
+
+  Block(int device, cudaStream_t stream, size_t size, uintptr_t ptr=0) :
+      device(device), stream(stream), size(size), ptr(ptr), allocated(0),
+      prev(NULL), next(NULL) { }
+};
+
+static bool BlockComparator(const Block* a, const Block* b)
+{
+  if (a->device != b->device) {
+    return a->device < b->device;
+  }
+  if (a->stream != b->stream) {
+    return (uintptr_t)a->stream < (uintptr_t)b->stream;
+  }
+  if (a->size != b->size) {
+    return a->size < b->size;
+  }
+  return a->ptr < b->ptr;
+}
+
+struct THCCachingAllocator
```

Comment:
After discussing with @wickedfoo, I've moved THCCachingAllocator outside the anonymous namespace, but did not add it to the header file since it contains C++